### Arguments Passing
This section covers some advanced topics related to Arguments Passing in
Python Functions.

### Mutable Default Parameter Values

Things can get weird if you specify a default parameter value that is a mutable object. Consider this Python function definition:

In [1]:
def f(my_list=[]):
    my_list.append('###')
    return my_list

In [2]:
f(['foo', 'bar', 'baz'])


['foo', 'bar', 'baz', '###']

f() takes a single list parameter, appends the string '###' to the end of the list, and returns the result:

Now consider this case:

In [3]:
f()

['###']

The above return value is as expected since the default value for the parameter  `my_list`  is empty list

what will happen if f() is called without any parameters a second and a third time? Let’s see:

In [4]:
f()


['###', '###']

In [5]:
f()


['###', '###', '###']

Unexpected behaviour right !!! In Python, default parameter values are defined only once when the function is defined (that is, when the def statement is executed). The default value isn’t re-defined each time the function is called. Thus, each time you call f() without a parameter, you’re performing .append() on the same list.

As a work around, consider using a default argument value that signals no argument has been specified. Most any value would work, but None is a common choice. When the sentinel value indicates no argument is given, create a new empty list inside the function:

In [6]:
def f(my_list=None):
    if my_list is None:
        my_list = []
    my_list.append('###')
    return my_list

In [7]:
f()

['###']

In [8]:
f()

['###']

In [9]:
f(['foo', 'bar', 'baz'])

['foo', 'bar', 'baz', '###']

### Pass by value vs Pass by reference in Python

As we saw in the lecture parameters in Python are nether passed by pass by value or pass by reference
Then how? we can clarify this with an example:


In [1]:
def f(fx):
    print('fx =', fx, '/ id(fx) = ', id(fx))
    fx = 10
    print('fx =', fx, '/ id(fx) = ', id(fx))


x = 5
print('x =', x, '/ id(x) = ', id(x))
f(x)
print('x =', x, '/ id(x) = ', id(x))

x = 5 / id(x) =  9788736
fx = 5 / id(fx) =  9788736
fx = 10 / id(fx) =  9788896
x = 5 / id(x) =  9788736


When `f()` first starts, `fx` and `x` both point to the same object, whose `id()`
 is 9788736. After `f()` executes the statement `fx = 10` , `fx` points to a
  different object whose `id()` is 9788896.
 The connection to the original object in the calling environment is lost.

 Argument passing in Python is somewhat of a hybrid between pass-by-value and pass-by-reference. What gets passed to the function is a reference to an object, but the reference is passed by value.



The key takeaway here is that a Python function can’t change the value of an argument by reassigning the corresponding parameter to something else. The following example demonstrates this:



In [4]:
def f(x):
    x = 'foo'

for i in (
        40,
        dict(foo=1, bar=2),
        {1, 2, 3},
        'bar',
        ['foo', 'bar', 'baz']):
    f(i)
    print(i)

40
{'foo': 1, 'bar': 2}
{1, 2, 3}
bar
['foo', 'bar', 'baz']


Here, objects of type int, dict, set, str, and list are passed to f() as arguments. f() tries to assign each to the string object 'foo', but as you can see, once back in the calling environment, they are all unchanged. As soon as f() executes the assignment x = 'foo', the reference is rebound, and the connection to the original object is lost.


#### Some Exception

What happens here?

In [1]:
def f(x):
    x[0] = '---'


my_list = ['foo', 'bar', 'baz', 'qux']

f(my_list)
my_list


['---', 'bar', 'baz', 'qux']

In this case, the argument to `f()` is a list. When `f()` is called,
 a reference to `my_list` is passed. You’ve already seen that `f()` can’t reassign `my_list`
  wholesale. If `x` were assigned to something else,
   then it would be bound to a different object,
    and the connection to `my_list` would be lost.

However, `f()` can use the reference to make modifications inside `my_list`.
Here, `f()` has modified the first element.
You can see that once the function returns,
`my_list` has, in fact, been changed in the calling environment.
The same concept applies to a dictionary


### Argument Passing Summary
Argument passing in Python can be summarized as follows.
**Passing an immutable object**, like an `int`, `str`, `tuple`, or `frozenset`,
 to a Python function acts like `pass-by-value`.
  **The function can’t modify the object in the calling environment**.

**Passing a mutable object** such as a `list`, `dict`, or `set` acts
 somewhat—but not exactly—like pass-by-reference.
  **The function can’t reassign the object wholesale**,
   but it can change items in place within the object,
   and these changes will be reflected in the calling environment.